### PR TITLE
Reduce memory/cache overhead from over loader processing

### PR DESCRIPTION
**Pull request by ijjk**
In `v14.0.2-canary.1` users started noticing more memory errors especially with many edge runtime configured pages. After investigation it seems this can be related to the additional transpiling we configured in https://github.com/vercel/next.js/pull/59569 and https://github.com/vercel/next.js/pull/57784

To help alleviate this we are updating the default swc loader to have an additional check to see if no special features such as `next/font`, `next/dynamic`, or `use server/client` directives are present and then no-oping in the loader to avoid additional overhead for a majority of modules. 

For monitoring regressions our `stats-app` has been updated with repeated edge-ssr routes to hopefully help us keep an eye on memory or cache size issues. 

x-ref: NEXT-2430
x-ref: NEXT-2395
x-ref: NEXT-2299
x-ref: NEXT-2324
x-ref: NEXT-2373

Closes NEXT-2479

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: ba6a0f76e071a4cffd192cf60db307b47d558630

`pnpm test-dev test/development/acceptance-app/ReactRefreshLogBox-builtins.test.ts`
- ReactRefreshLogBox app default > Node.js builtins

<details>
<summary>Expand output</summary>

  ● ReactRefreshLogBox app default › Node.js builtins

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `ReactRefreshLogBox app default Node.js builtins 1`

    - Snapshot  - 1
    + Received  + 1

    @@ -1,6 +1,6 @@
    - ./node_modules/my-package/index.js:1:13
    + ./node_modules/my-package/index.js:1:1
      Module not found: Can't resolve 'dns'

      https://nextjs.org/docs/messages/module-not-found

      Import trace for requested module:

      62 |       `)
      63 |     } else {
    > 64 |       expect(await session.getRedboxSource()).toMatchInlineSnapshot(`
         |                                               ^
      65 |         "./node_modules/my-package/index.js:1:13
      66 |         Module not found: Can't resolve 'dns'
      67 |

      at Object.toMatchInlineSnapshot (development/acceptance-app/ReactRefreshLogBox-builtins.test.ts:64:47)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js update/exclude-conditions | Change |
| - | - | - | - |
| buildDuration | 12.1s | 11.7s | N/A |
| buildDurationCached | 6.6s | 5.1s | N/A |
| nodeModulesSize | 196 MB | 196 MB | ⚠️ +1.4 kB |
| nextStartRea..uration (ms) | 430ms | 406ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js update/exclude-conditions | Change |
| - | - | - | - |
| 3f784ff6-HASH.js gzip | 53.5 kB | 53.5 kB | ✓ |
| 423.HASH.js gzip | 185 B | 181 B | N/A |
| 68-HASH.js gzip | 30 kB | 30 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 238 B | 240 B | N/A |
| main-HASH.js gzip | 31.9 kB | 31.9 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 100 kB | 100 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js update/exclude-conditions | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js update/exclude-conditions | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 502 B | 501 B | N/A |
| css-HASH.js gzip | 320 B | 322 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 256 B | N/A |
| head-HASH.js gzip | 350 B | 349 B | N/A |
| hooks-HASH.js gzip | 368 B | 369 B | N/A |
| image-HASH.js gzip | 4.2 kB | 4.2 kB | N/A |
| index-HASH.js gzip | 257 B | 256 B | N/A |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 310 B | 311 B | N/A |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 306 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 106 B | 106 B | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js update/exclude-conditions | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js update/exclude-conditions | Change |
| - | - | - | - |
| index.html gzip | 527 B | 527 B | ✓ |
| link.html gzip | 541 B | 539 B | N/A |
| withRouter.html gzip | 523 B | 522 B | N/A |
| Overall change | 527 B | 527 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js update/exclude-conditions | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94.4 kB | 94.4 kB | N/A |
| page.js gzip | 150 kB | 150 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js update/exclude-conditions | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 620 B | 624 B | N/A |
| middleware-r..fest.js gzip | 151 B | 149 B | N/A |
| middleware.js gzip | 47.4 kB | 47.4 kB | ✓ |
| edge-runtime..pack.js gzip | 1.94 kB | 1.94 kB | ✓ |
| Overall change | 49.3 kB | 49.3 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js update/exclude-conditions | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 166 kB | 166 kB | ✓ |
| app-page-exp..prod.js gzip | 95.5 kB | 95.5 kB | ✓ |
| app-page-tur..prod.js gzip | 97.2 kB | 97.2 kB | ✓ |
| app-page-tur..prod.js gzip | 91.7 kB | 91.7 kB | ✓ |
| app-page.run...dev.js gzip | 136 kB | 136 kB | ✓ |
| app-page.run..prod.js gzip | 90.2 kB | 90.2 kB | ✓ |
| app-route-ex...dev.js gzip | 22 kB | 22 kB | ✓ |
| app-route-ex..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.6 kB | 14.6 kB | ✓ |
| app-route.ru...dev.js gzip | 21.7 kB | 21.7 kB | ✓ |
| app-route.ru..prod.js gzip | 14.6 kB | 14.6 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.43 kB | 9.43 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.7 kB | 9.7 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.43 kB | 9.43 kB | ✓ |
| pages-turbo...prod.js gzip | 22.1 kB | 22.1 kB | ✓ |
| pages.runtim...dev.js gzip | 22.7 kB | 22.7 kB | ✓ |
| pages.runtim..prod.js gzip | 22.1 kB | 22.1 kB | ✓ |
| server.runti..prod.js gzip | 49.9 kB | 49.9 kB | ✓ |
| Overall change | 925 kB | 925 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js update/exclude-conditions | Change |
| - | - | - | - |
| 0.pack gzip | 1.63 MB | 1.54 MB | N/A |
| index.pack gzip | 61.6 kB | 61.7 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

```diff
failed to diff
```
</details>
<details>
<summary>Diff for <strong>68-HASH.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: dea9e85f340a2d3aac3edfe2bd8c762f88198cef

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 6383547e1a3fd998474a55f2f7f12c05ec1eea03

`pnpm test-dev test/development/acceptance-app/ReactRefreshLogBox-builtins.test.ts`
- ReactRefreshLogBox app default > Node.js builtins

<details>
<summary>Expand output</summary>

  ● ReactRefreshLogBox app default › Node.js builtins

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `ReactRefreshLogBox app default Node.js builtins 1`

    - Snapshot  - 1
    + Received  + 1

    @@ -1,6 +1,6 @@
    - ./node_modules/my-package/index.js:1:13
    + ./node_modules/my-package/index.js:1:1
      Module not found: Can't resolve 'dns'

      https://nextjs.org/docs/messages/module-not-found

      Import trace for requested module:

      62 |       `)
      63 |     } else {
    > 64 |       expect(await session.getRedboxSource()).toMatchInlineSnapshot(`
         |                                               ^
      65 |         "./node_modules/my-package/index.js:1:13
      66 |         Module not found: Can't resolve 'dns'
      67 |

      at Object.toMatchInlineSnapshot (development/acceptance-app/ReactRefreshLogBox-builtins.test.ts:64:47)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: test/.stats-app/app/edge-ssr-repeated-1/page.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import Image from 'next/image'
```
Comment (ijjk) [inline test/.stats-app/app/edge-ssr-repeated-1/page.js:1]: These are intentional to stress test compiling numerous edge app routes.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: becc714b3384060dd605ab506871c20a7239334a

`pnpm test-start test/e2e/prerender.test.ts`
- Prerender > should on-demand revalidate for fallback: blocking with onlyGenerated if generated

<details>
<summary>Expand output</summary>

  ● Prerender › should on-demand revalidate for fallback: blocking with onlyGenerated if generated

    expect(received).toMatch(expected)

    Expected pattern: /(HIT|STALE)/
    Received string:  "MISS"

      2294 |
      2295 |         expect(initialTime).toBe($2('#time').text())
    > 2296 |         expect(res2.headers.get('x-nextjs-cache')).toMatch(/(HIT|STALE)/)
           |                                                    ^
      2297 |
      2298 |         const res3 = await fetchViaHTTP(
      2299 |           next.url,

      at Object.toMatch (e2e/prerender.test.ts:2296:52)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: c5eee6ac05cd6a1aa4f16333582f66fdd6590116

`pnpm test test/integration/prerender-export/test/index.test.js`

<details>
<summary>Expand output</summary>



  ● Test suite failed to run

    thrown: "Exceeded timeout of 60000 ms for a hook.
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."

      217 |       })
      218 |
    > 219 |       afterAll(async () => {
          |       ^
      220 |         if (app) {
      221 |           await stopApp(app)
      222 |         }

      at afterAll (integration/prerender-export/test/index.test.js:219:7)
      at describe (integration/prerender-export/test/index.test.js:209:5)
      at integration/prerender-export/test/index.test.js:208:52
      at Object.describe (integration/prerender-export/test/index.test.js:207:1)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: df14ec52ac5c4ab959893032569d37e603fb88af

`pnpm test-start test/e2e/app-dir/app-client-cache/client-cache.test.ts`
- app dir client cache semantics > prefetch={true} > should prefetch again after 5 mins if the link is visible again

<details>
<summary>Expand output</summary>

  ● app dir client cache semantics › prefetch={true} › should prefetch again after 5 mins if the link is visible again

    TIMED OUT: success

    fail

    undefined

      636 |
      637 |   if (hardError) {
    > 638 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      639 |   }
      640 |   return false
      641 | }

      at check (lib/next-test-utils.ts:638:11)
      at Object.<anonymous> (e2e/app-dir/app-client-cache/client-cache.test.ts:176:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: df14ec52ac5c4ab959893032569d37e603fb88af

`pnpm test test/integration/preload-viewport/test/index.test.js`
- Prefetching Links in viewport > production mode > should prefetch with non-bot UA

<details>
<summary>Expand output</summary>

  ● Prefetching Links in viewport › production mode › should prefetch with non-bot UA

    expect(received).toHaveLength(expected)

    Expected length: 1
    Received length: 0
    Received array:  []

      129 |         )
      130 |         const links = await browser.elementsByCss('link[rel=prefetch]')
    > 131 |         expect(links).toHaveLength(1)
          |                       ^
      132 |       } finally {
      133 |         if (browser) await browser.close()
      134 |       }

      at Object.toHaveLength (integration/preload-viewport/test/index.test.js:131:23)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 9eceee87b030adb0d9cf7e6da715cb7d54660e06

`pnpm test-dev test/e2e/app-dir/interception-route-prefetch-cache/interception-route-prefetch-cache.test.ts` (PPR)
- interception-route-prefetch-cache > runtime = edge > should render the correct interception when two distinct layouts share the same path structure

<details>
<summary>Expand output</summary>

  ● interception-route-prefetch-cache › runtime = edge › should render the correct interception when two distinct layouts share the same path structure

    TIMED OUT: /Intercepted on Foo Page/



    TimeoutError: page.waitForSelector: Timeout 60000ms exceeded.
    =========================== logs ===========================
    waiting for locator('#slot')
    ============================================================

      636 |
      637 |   if (hardError) {
    > 638 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      639 |   }
      640 |   return false
      641 | }

      at check (lib/next-test-utils.ts:638:11)
      at Object.<anonymous> (e2e/app-dir/interception-route-prefetch-cache/interception-route-prefetch-cache.test.ts:21:7)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 9eceee87b030adb0d9cf7e6da715cb7d54660e06

`pnpm test-start test/e2e/app-dir/navigation/navigation.test.ts` (PPR)
- app dir - navigation > navigation between pages and app > should not continously initiate a mpa navigation to the same URL when router state changes

<details>
<summary>Expand output</summary>

  ● app dir - navigation › navigation between pages and app › should not continously initiate a mpa navigation to the same URL when router state changes

    page.waitForSelector: Timeout 60000ms exceeded.
    =========================== logs ===========================
    waiting for locator('#link-to-slow-page')
      locator resolved to visible <a href="/slow-page" id="link-to-slow-page">To /slow-page</a>
    ============================================================

      422 |     return this.chain(() => {
      423 |       return page
    > 424 |         .waitForSelector(selector, { timeout, state: 'attached' })
          |          ^
      425 |         .then(async (el) => {
      426 |           // it seems selenium waits longer and tests rely on this behavior
      427 |           // so we wait for the load event fire before returning

      at waitForSelector (lib/browsers/playwright.ts:424:10)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: packages/next/src/build/webpack/loaders/next-swc-loader.ts [hunk 2]
```diff
@@ -58,6 +84,20 @@ async function loaderTransform(
...
+    filename,
+    loaderOptions.transpilePackages || []
+  )
+
+  if (shouldMaybeExclude) {
+    if (!source) {
+      throw new Error(`Invariant might be excluded but missing source`)
+    }
+
+    if (!FORCE_TRANSPILE_CONDITIONS.some((item) => source.includes(item))) {
```
[the remark this answers is not in the document]
Comment (ijjk) [reply packages/next/src/build/webpack/loaders/next-swc-loader.ts]: Yea might be better, will update

**Comment by ijjk** [comment]
## Failing test suites 
Commit: c83a4bc8f9a2a0d6a9cb07be0c39c656146e7db5

`pnpm test test/integration/prerender-export/test/index.test.js`

<details>
<summary>Expand output</summary>



  ● Test suite failed to run

    thrown: "Exceeded timeout of 60000 ms for a hook.
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."

      217 |       })
      218 |
    > 219 |       afterAll(async () => {
          |       ^
      220 |         if (app) {
      221 |           await stopApp(app)
      222 |         }

      at afterAll (integration/prerender-export/test/index.test.js:219:7)
      at describe (integration/prerender-export/test/index.test.js:209:5)
      at integration/prerender-export/test/index.test.js:208:52
      at Object.describe (integration/prerender-export/test/index.test.js:207:1)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by huozhi — APPROVED** [review]

**Pull request merged**